### PR TITLE
Wallet Sync Speedup

### LINKED ISSUE #1103: Wallet sync speedup
So let's say we have a wallet that is 1000 blocks behind the block store tip. WalletSyncManager will get those block one by one in order to update the wallet. 
And we know that dbreeze stores stuff in sorted order and range selects are very fast (or so the official documentation says). So why don't we just use that and get those missing blocks in just large chanks (lets say, of 100 blocks) instead of getting one block at a time? 
So I propose: 
1) Avoid using block store cache in the WalletSyncManager (that doesn't make much sense to cache old blocks that we will never use after).
2) Adding `GetBlocksAsync` to the `IBlockRepository` and implementing it.
3) Making `WalletSyncManager` utilize that new `GetBlocksAsync` method when wallet tip is behind the block store cache tip by more than 1 block.

### PR DESCRIPTION
**Pull request by ferdeen**
Fixes #1103 

The change to allow the Wallet to move forward while syncing is based on the[ Producer Consumer DataFlow pattern](https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/how-to-implement-a-producer-consumer-dataflow-pattern).  Which has been implemented locally in the `WalletSyncManager` class.

Tests impacted : 

image

Syncing against all for Daemons (Bitcoin, Stratis - MainNet and TestNet).

image

The syncing is currently running and I will be keeping an eye on any issues, but thought it would be best to get the PR out there for any comments on the solution - and hit the sprint deadline.

Cheers,
Ferdeen

### CODE REVIEW

**Comment by monsieurleberre** [comment]
Starting the review now :)

File: src/Stratis.Bitcoin.Features.BlockStore/BlockStoreCache.cs [hunk 1]
```diff
@@ -79,6 +88,44 @@ public virtual BlockStoreCachePerformanceCounter BlockStoreCachePerformanceCount
             return block;
         }
 
+        /// <inheritdoc />
+        public async Task<List<Block>> GetBlockAsync(List<uint256> blockids)
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin.Features.BlockStore/BlockStoreCache.cs]: GetBlock*s*Async ?

Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.BlockStore/BlockStoreCache.cs]: done

File: src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs [hunk 2]
```diff
@@ -96,158 +119,216 @@ public void Stop()
...
-                    this.walletTip = fork;
+                        this.walletManager.RemoveBlocks(fork);
+                        this.walletTip = fork;
 
-                    this.logger.LogTrace("Wallet tip set to '{0}'.", this.walletTip);
+                        this.logger.LogTrace("Wallet tip set to '{0}'.", this.walletTip);
+                    }
+
+                    // The new tip can be ahead or behind the wallet.
+                    // If the new tip is ahead we try to bring the wallet up to the new tip.
```
Comment (monsieurleberre) [inline src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: Was the change form > to < intended ?

Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: changed to <=

**Comment by ferdeen** [comment]
@noescape00 the current blockstorecache is working fine - the wallet height is advancing as normal.  

I am not too sure why you need to change it now.

When I made the change locally `IBlockStore.GetBlockAsync()` is not implemented - so I have undone the change.

https://github.com/stratisproject/StratisBitcoinFullNode/blob/16f602fb1c08082ae2d86ee6b2c7c8c79dbbf979/src/Stratis.Bitcoin.Features.BlockStore/BlockStoreQueue.cs#L170-L174

Also this method is used in ConsensusManager - which if not implemented, doesn't seem right.

https://github.com/stratisproject/StratisBitcoinFullNode/blob/a238bc18bf97bfab93b8e936cd92dc8ed55d53f2/src/Stratis.Bitcoin/Consensus/ConsensusManager.cs#L1043

**Comment by noescape00** [comment]
>@noescape00 the current blockstorecache is working fine - the wallet height is advancing as normal.
>I am not too sure why you need to change it now.

It does. However after the activation we will have only 2 block providers- ConsensusManager and BlockStoreQueue. CHT will do the caching of the last 100 connected blocks. 

So any change to BlockStoreCache will become obsolete after activation is merged.

**Comment by ferdeen** [comment]
@noescape00 Happy to do this.  However this is turning into something different.  Maybe we need a new card for this related to the activation work ?

@bokobza what do you think ?

**Comment by ferdeen** [comment]
Had a chat with @noescape00 on Slack and the change is a one liner -

 `return this.blockRepository.GetAsync(blockHash);`

However, after adding this after a cycle the wallet doesn't advance.  I'll check the logging to see why this is the case.

Cheers

File: src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs (hunk 2 above)
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/WalletSyncManager.cs]: Hi @monsieurleberre,

When I test wallet syncing the logic changed from > to < by mistake.  

I have reverted the change and testing to make sure the wallet sync's in the correct manner.

Cheers,

Ferdeen

File: src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs [hunk 3]
```diff
@@ -852,7 +852,7 @@ public void ProcessBlock(Block block, ChainedHeader chainedHeader)
                 if (chainedHeader.Height > current.Height)
                 {
                     this.logger.LogTrace("(-)[BLOCK_TOO_FAR]");
-                    throw new WalletException("block too far in the future has arrived to the wallet");
+                    return;
```
[the remark this answers is not in the document]
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: ok

[the remark this answers is not in the document]
Comment (ferdeen) [reply src/Stratis.Bitcoin.Features.Wallet/WalletManager.cs]: Carry on as we now store any blocks that come in from consensus in the local concurrent queue.

**Comment by ferdeen** [comment]
New PR (#1869) created that is based on today's master.

**Pull request closed without merging**